src/Nethermind/Nethermind.Taiko/Rpc/SurgeGasPriceOracle.cs

```diff
@@ -73,11 +75,22 @@ public override UInt256 GetGasPriceEstimate()
 
         UInt256 proofPostingCost = _surgeConfig.ProofPostingGas * UInt256.Max(l1BaseFee, l1AverageBaseFee);
 
-        UInt256 gasPriceEstimate = (minProposingCost + proofPostingCost + _surgeConfig.ProvingCostPerL2Batch) / _surgeConfig.L2GasPerL2Batch;
+        // Record the current block's gas usage and compute the average
+        _gasUsageBuffer.Add((ulong)headBlock.GasUsed);
+        UInt256 averageGasUsage = _gasUsageBuffer.Average;
+
+        if (averageGasUsage == UInt256.Zero)
+        {
+            if (_logger.IsTrace) _logger.Trace("[SurgeGasPriceOracle] Average gas usage is zero, using fallback gas price");
+            return FallbackGasPrice();
+        }
+
+        UInt256 gasPriceEstimate = (minProposingCost + proofPostingCost + _surgeConfig.ProvingCostPerL2Batch) / averageGasUsage;
```

Comment:
need to do this per batch not per block. batch contains multiple blocks. you need to read it directly from the smart contract. Check with @AnshuJalan